### PR TITLE
[Runtimes] Limit function's service accounts based on project configuration

### PR DESCRIPTION
**Pull request by theSaarco**
Up until now there was no mechanism to limit the service account that a user can set on a function (using `function.spec.service_account = '<something>'`. This is a potential security issue, for example if users are utilizing Google workload identity that maps service accounts to Google service accounts.

This PR adds the following mechanism to limit the service accounts:
Use 2 internal project secrets (per project):
* `mlrun.service-accounts.allowed` - a list of allowed service accounts, separated by `,`. For example - `sa1, sa2, sa3`
* `mlrun.service-accounts.default` - a default service account per project. This will be automatically assigned to every function deployed from that function, unless the user manually set a different SA

The logic is as follows:

1. If the allowed SAs list is empty (secret does not exist) - the logic is exactly like today, i.e. there's no restriction on the SAs that the user can assign to a function.
2. If a default SA is provided and a function does not have an SA assigned, then set the default SA to the function. If the function already has a SA, keep it.
3. If the allowed SA list is not empty, then validate that the function's SA exists in the list of allowed SAs. If it is not, fail the submitting/deploying of the function.

It's valid to have a default SA without an allowed list. It's also valid to have an allowed list without a default SA.

Currently this PR only adds the validation side, it does not add MLRun APIs to set the allowed list or the default SA. This will be done in the future once we make a decision on the permissions needed for this API and how this should be managed. This can be utilized by externally creating the relevant secret keys per project (it cannot be done through MLRun's project-secret mechanism, since the secrets are internal secrets and MLRun will block it).

Another fix in this PR - up until now the `func.spec.service_account` property completely ignored for nuclio functions. This PR fixes it.

### CODE REVIEW

**Review by Hedingber — CHANGES_REQUESTED** [review]
I think you're missing the implementation for Dask (see the call to `ensure_function_has_auth_set` from `_start_function`)

File: tests/api/api/test_submit.py [hunk 1]
```diff
@@ -150,6 +149,67 @@ def test_submit_job_ensure_function_has_auth_set(
...
+        ] = base64.b64encode(default_service_account.encode()).decode("utf-8")
+    if allowed_service_accounts:
+        allowed_str = ",".join(allowed_service_accounts)
+        secrets[
+            mlrun.api.crud.secrets.Secrets().generate_service_account_secret_key(
+                "allowed"
+            )
+        ] = base64.b64encode(allowed_str.encode()).decode("utf-8")
+
+    get_k8s()._get_project_secrets_raw_data = unittest.mock.Mock(return_value=secrets)
```
[the remark this answers is not in the document]
Comment (theSaarco) [reply tests/api/api/test_submit.py]: Using it.

File: tests/api/api/test_utils.py [hunk 2]
```diff
@@ -1,12 +1,25 @@
...
+
+
+@pytest.fixture(autouse=True)
+def mock_secrets_call():
+    orig_function = get_k8s()._get_project_secrets_raw_data
+    get_k8s()._get_project_secrets_raw_data = unittest.mock.Mock(return_value={})
+
+    yield
+
+    get_k8s()._get_project_secrets_raw_data = orig_function
```
[the remark this answers is not in the document]
Comment (theSaarco) [reply tests/api/api/test_utils.py]: yep

File: tests/api/utils/test_scheduler.py [hunk 3]
```diff
@@ -58,6 +59,16 @@ async def do_nothing():
     pass
 
 
+@pytest.fixture(autouse=True)
+def mock_secrets_call():
```
[the remark this answers is not in the document]
Comment (theSaarco) [reply tests/api/utils/test_scheduler.py]: right

**Review by Hedingber — APPROVED** [review]

**Pull request merged**